### DISCUSSION TITLE
Better TypeScript Support

### DISCUSSION THREAD

**Discussion by SuperSonicHub1** [tinyplex/tinybase#33] [Ideas]
I'm using TinyBase in a TypeScript web app and would like to be able to statically enforce my schema instead of using TypeScript instead of at runtime using `Table.setSchema`. I think it could work something like this:

```
type TableSchema = {
  pets: {
    species: string,
    sold: boolean,
  },
}

const store = createStore<TableSchema>();
```
This will allow me to reuse existing types with TinyBase and not have TypeScript be mad at me. :)
If this is already possible, please let me know.

**Comment by jamesgpearce** [comment]
Hey @SuperSonicHub1! You should take a look at the v2.2 release (currently in beta, but set to roll today). It's not quite the generic-based implementation you suggest, but it can can (either programmatically or on the command line) take the schema and generate a typed wrapper with domain-specific methods etc.

Take a look at https://beta.tinybase.org/guides/developer-tools/generating-apis/ and https://beta.tinybase.org/guides/developer-tools/command-line/ and let me know what you think. (In a couple of hours, these URLs will also work without the beta subdomain ;) )

**Comment by SuperSonicHub1** [comment]
Great! I guess next steps for me are to take my TypeScript schemas and convert them to TinyBase ones.

Thanks for working on TinyBase; it's been a lot of fun writing a offline-first Mastodon client that implements the [relational philosopy of Riffle](https://riffle.systems/essays/prelude/) using it! I should have a 0.1.0 of the app done in the coming weeks; I'll make sure to share it here as I think it might be helpful for you guys seeing as I'm likely one of the first devs to write a full app with TinyBase (if you do know of any others, please let me know!).

**Reply by jamesgpearce** [reply]
Awesome! Let me know - would love to see that.

**Comment by erwinkn** [comment]
Another idea for better TypeScript support: infer types for tables, rows, and cells from a schema and use those to strongly type all the functions of the library, directly in TypeScript. I'm thinking of something in the style of [`zod`](https://github.com/colinhacks/zod) and its [`z.infer<typeof schema>`](https://zod.dev/?id=type-inference).

Example:
```ts
const store = createStore().setSchema({
  pets: {
    species: { type: "string" },
    sold: { type: "boolean" },
  },
});

// Should be a type error
store.getTable("pet");

// Should have type: `Record<string, { species: string; sold: boolean; }>`
const table = store.getTable("pets");

// Should have type `{ species: string; sold: boolean; }`
const row = table["felix"];
```

It has the advantage of not requiring any code generation. I suspect this approach could also accept a generic argument to `createStore`.

If this is of interest, I'd be happy to work on it, I've implemented similar TypeScript tricks in the past.

We're planning on using TinyBase in our app for offline support and cross-tab sync & this feature would be very helfpul.

**Reply by SuperSonicHub1** [reply]
This sounds like a great idea! I wonder if we could bring in Zod as a dependency to implement this instead of rolling our solution.

**Reply by erwinkn** [reply]
Ended up starting on an implementation with the current API. It currently covers getters, setters, and most listeners for the `Store` interface.

Listeners are especially interesting (and challenging!), but it's possible to get full typesafety, even with wildcards or the `getCellChange` function. Well, *nearly* full, the demo shows the one case I've found where there is a gap in type inference.

Here's a demo: https://codesandbox.io/s/tinybase-types-demo-m9l2ui?file=/src/index.ts

**Reply by SuperSonicHub1** [reply]
Awesome!

**Reply by WonderPanda** [reply]
I definitely feel like this effort will be a game changer for TinyBase adoption. I'd be very keen to see this implemented into the core so that we get better TS ergonomics up front without needing a codegen step

**Reply by jamesgpearce** [reply]
I'm supportive! I'll just need to think about how to explain the difference between this and the codegen API approach. But if you think it has widespread value, let's do it! Happy to review and start integrating when it's ready.

**Reply by WonderPanda** [reply]
@jamesgpearce There's been a huge renaissance inside the TypeScript community when it comes to the power of inferring types for incredible developer experiences. [Zod](https://github.com/colinhacks/zod) and [tRPC](https://trpc.io/) have been pushing the boundaries here.

If we can get TinyBase to the point where we can get inferred typesafety without leaning on codegen for the majority of the API including some of the more advanced pieces of TinyQL I really feel like the appeal of the library will be massive

@erwinkn Are you interested in opening up a PR with these improved type definitions?

**Reply by erwinkn** [reply]
Absolutely! I'd like to get the types working for the `store` and `queries` modules first though, as I think those are the most challenging. I'm also working on a fully typesafe query builder for TinyQL and it would be nice to make sure the core library types make it practical to write such a query builder.

I have been working on them in a separate repo, just because it was easier to get started: https://github.com/erwinkn/tinybase-types

I think the `store` module is complete, although I haven't extensively tested it.

**Comment by jamesgpearce** [comment]
Hey @erwinkn! Just checking in to see how this is going :) 

Also wanted to make sure you were keeping an eye on the `keyvalues` branch, where I am adding first-class support for key-values (basically every Store can have a flat map of Cell-like 'Value' entries keyed by Id). I think this should be easy enough to type and it also has schema and tools support.

**Comment by jamesgpearce** [comment]
Hey @erwinkn! Just checking in to see how this is going :) 

Also wanted to make sure you were keeping an eye on the `keyvalues` branch, where I am adding first-class support for key-values (basically every Store can have a flat map of Cell-like 'Value' entries keyed by Id). I think this should be easy enough to type and it also has schema and tools support.

**Reply by erwinkn** [reply]
Hey James! Sorry, I missed this notification.
My workload is easing up, so I'll be able to get back onto `tinybase-types` this weekend. Once I have the typesafe query builder, I'll transfer them into a pull request for this repo :)

**Reply by jamesgpearce** [reply]
no problemo!

**Comment by jamesgpearce** [comment]
Keep an eye on #50, the beta branch, and https://twitter.com/jamespearce/status/1631406755082117121 - this will be coming in v3.1, not too differently to what was asked for here!

**Comment by jamesgpearce** [comment]
Please kick the tires on https://github.com/tinyplex/tinybase/releases/tag/v3.1.0-beta.3

(So far I couldn't do this without some thin code generation ('refinement') because the schema can be replaced and removed at runtime. I suspect that to fulfill the dream I'll need a core Store API change, so maybe a v4.0 sort of thing.)
